src/shader/shaderLightSampling.js

```diff
@@ -73,14 +110,103 @@ LightSampleRec randomRectAreaLightSample( Light light, vec3 rayOrigin ) {
 
 }
 
+LightSampleRec randomCircularAreaLightSample( Light light, vec3 rayOrigin ) {
+
+	LightSampleRec lightSampleRec;
+	lightSampleRec.hit = true;
+
+	lightSampleRec.emission = light.color * light.intensity;
+
+	float r = 2.0 * sqrt( rand() );
+	float theta = rand() * 2.0 * PI;
+	float x = r * cos( theta );
+	float y = r * sin( theta );
+
+	vec3 randomPos = light.position + light.u * x + light.v * y;
+	vec3 toLight = randomPos - rayOrigin;
+	float lightDistSq = dot( toLight, toLight );
+	lightSampleRec.dist = sqrt( lightDistSq );
+
+	vec3 direction = toLight / lightSampleRec.dist;
+	lightSampleRec.direction = direction;
+
+	vec3 lightNormal = normalize( cross( light.u, light.v ) );
+	lightSampleRec.pdf = lightDistSq / ( light.area * dot( direction, lightNormal ) );
+
+	return lightSampleRec;
+
+}
+
+LightSampleRec randomSpotLightSample( SpotLight spotLight, sampler2DArray iesProfiles, vec3 rayOrigin ) {
+
+	LightSampleRec lightSampleRec;
+	lightSampleRec.hit = true;
+
+	float r = 2.0 * spotLight.radius * sqrt( rand() );
```

Comment:
I haven't tried to look for any issues with this visually but is it correct that this should be multiplied by 2.0, as well? Assuming that the radius of the spot lights directly represents the radius of a disc then shouldn't this be multiplied by 1.0 instead of 2.0 since theta is sampled from [0, 2 PI]?

Reply:
This slipped through when experimenting, removing in the next push.